freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/UserSyncService.java

```diff
@@ -151,29 +164,8 @@ private void asyncSynchronizeUsers(
 
             Map<String, Future<SyncStatusDetail>> statusFutures = stacks.stream()
                     .collect(Collectors.toMap(Stack::getEnvironmentCrn,
-                            stack -> asyncTaskExecutor.submit(() -> {
-                                MDCBuilder.buildMdcContext(stack);
-                                String envCrn = stack.getEnvironmentCrn();
-                                SyncStatusDetail statusDetail =
-                                        synchronizeStack(stack, envToUmsStateMap.get(stack.getEnvironmentCrn()), userCrnFilter, machineUserCrnFilter);
-                                switch (statusDetail.getStatus()) {
-                                    case COMPLETED:
-                                        success.add(new SuccessDetails(envCrn));
-                                        if (umsEventGenerationIdsJson != null) {
-                                            UserSyncStatus userSyncStatus = userSyncStatusService.getOrCreateForStack(stack);
-                                            userSyncStatus.setUmsEventGenerationIds(umsEventGenerationIdsJson);
-                                            userSyncStatusService.save(userSyncStatus);
-                                        }
-                                        break;
-                                    case FAILED:
-                                        failure.add(new FailureDetails(envCrn, statusDetail.getDetails()));
-                                        break;
-                                    default:
-                                        failure.add(new FailureDetails(envCrn, "Unknown status"));
-                                        break;
-                                }
-                                return statusDetail;
-                            })));
+                            stack -> asyncSynchronizeStack(stack, envToUmsStateMap.get(stack.getEnvironmentCrn()),
+                                    umsEventGenerationIds, d -> success.add(d), d -> failure.add(d), fullSync)));
```

Comment:
you might use method reference for these: `d -> success.add(d), d -> failure.add(d)`

Reply:
also this `ConcurrentLinkedQueue` stuff doesn't seem right to me. You already collect the results into `statusFutures` and iterate over them. Is it possible to put there the necessary info and create the `successDetails` and `failureDetails` from those?

Reply:
fixed